src/github/pullRequestManager.ts

```diff
@@ -647,31 +639,66 @@ export class PullRequestManager implements vscode.Disposable {
-			for (const repository of this._githubRepositories) {
-				this._repositoryPageInformation.set(repository.remote.url.toString(), {
-					pullRequestPage: 1,
+		const queryId = type.toString() + (query || '');
+		for (const repository of this._githubRepositories) {
+			const remoteId = repository.remote.url.toString() + queryId;
+			if (!this._repositoryPageInformation.get(remoteId)) {
+				this._repositoryPageInformation.set(remoteId, {
+					pullRequestPage: 0,
 					hasMorePages: null
 				});
 			}
 		}
 
+		const pullRequestData: PullRequestData = { hasMorePages: false, pullRequests: [] };
+		const addPage = (page: PullRequestData | undefined) => {
+			if (page) {
+				pullRequestData.pullRequests = pullRequestData.pullRequests.concat(page.pullRequests);
+				pullRequestData.hasMorePages = page.hasMorePages;
+			}
+		};
+
 		const githubRepositories = this._githubRepositories.filter(repo => {
-			const info = this._repositoryPageInformation.get(repo.remote.url.toString());
-			return info && info.hasMorePages !== false;
+			const info = this._repositoryPageInformation.get(repo.remote.url.toString() + queryId);
+			// If we are reloading (fetchNextPage === false), don't filter out repos that are out of pages.
+			return info && (options.fetchNextPage === false || info.hasMorePages !== false);
 		});
 
 		for (let i = 0; i < githubRepositories.length; i++) {
 			const githubRepository = githubRepositories[i];
-			const pageInformation = this._repositoryPageInformation.get(githubRepository.remote.url.toString())!;
-			const pullRequestData = type === PRType.All
-				? await githubRepository.getAllPullRequests(pageInformation.pullRequestPage)
-				: await githubRepository.getPullRequestsForCategory(query || '', pageInformation.pullRequestPage);
+			const remoteId = githubRepository.remote.url.toString() + queryId;
+			const pageInformation = this._repositoryPageInformation.get(remoteId)!;
+
+			const fetchPage = async (pageNumber: number) =>
+				type === PRType.All
+					? await githubRepository.getAllPullRequests(pageNumber)
+					: await githubRepository.getPullRequestsForCategory(query || '', pageNumber);
+
+			if (options.fetchNextPage) {
+				pageInformation.pullRequestPage++;
+				addPage(await fetchPage(pageInformation.pullRequestPage));
+			} else {
+				if (pageInformation.pullRequestPage === 0) { pageInformation.pullRequestPage = 1; }
+				const pages = await Promise.all(
+					Array.from({ length: pageInformation.pullRequestPage }).map((_, j) => fetchPage(j + 1)));
+				pages.forEach(page => addPage(page));
+			}
+
+			pageInformation.hasMorePages = pullRequestData.hasMorePages;
+
+			const repoNumber = this._githubRepositories.findIndex(
+				repo => repo.remote.url.toString() === githubRepository.remote.url.toString()) + 1;
 
-			pageInformation.hasMorePages = !!pullRequestData && pullRequestData.hasMorePages;
-			pageInformation.pullRequestPage++;
+			this._numFetchedRepos = Math.max(this._numFetchedRepos, repoNumber);

```

Comment:
Can this be simplified to Math.max(this._numFetchedRepos, i)? I'm hitting an off-by-one problem in a repo that has `origin` with no prs and `upstream` with actual results, `repoNumber` gets set to 2

Reply:
Moved to a different strategy for keeping track of past fetches.